File: app/src/main/java/com/coinninja/coinkeeper/db/CoinKeeperOpenHelper.java
package com.coinninja.coinkeeper.db;

import com.coinninja.coinkeeper.di.interfaces.DBEncryption;

import org.greenrobot.greendao.database.Database;

import javax.inject.Inject;

public class CoinKeeperOpenHelper {
    public static final String TAG = CoinKeeperOpenHelper.class.getSimpleName();
    private final DatabaseOpenHelper databaseOpenHelper;
    private final DatabaseSecretProvider databaseSecretProvider;
    private final boolean withEncryption;


    @Inject
    CoinKeeperOpenHelper(DatabaseOpenHelper databaseOpenHelper, DatabaseSecretProvider databaseSecretProvider,
                         @DBEncryption boolean withEncryption) {
        this.databaseOpenHelper = databaseOpenHelper;
        this.databaseSecretProvider = databaseSecretProvider;
        this.withEncryption = withEncryption;
    }

    public Database getWritableDatabase() {
        if (withEncryption) {
            return getEncryptedWritableDb();
        } else {
            return databaseOpenHelper.getWritableDb();
        }
    }

    private Database getEncryptedWritableDb() {
        try {
            return databaseOpenHelper.getEncryptedWritableDb(databaseSecretProvider.getSecret());
        } catch (net.sqlcipher.database.SQLiteException ex) {
            return databaseOpenHelper.getEncryptedWritableDb(databaseSecretProvider.getDefault());
        }
    }


}

File: app/src/main/java/com/coinninja/coinkeeper/db/CoinKeeperOpenHelper.kt
package com.coinninja.coinkeeper.db

import app.dropbit.annotations.Mockable
import com.coinninja.coinkeeper.di.interfaces.DBEncryption
import net.sqlcipher.database.SQLiteException
import org.greenrobot.greendao.database.Database

@Mockable
class CoinKeeperOpenHelper internal constructor(
        internal val databaseOpenHelper: DatabaseOpenHelper,
        internal val databaseSecretProvider: DatabaseSecretProvider,
        internal val upgradeDBFormatStorage: UpgradeDBFormatStorage,
        @DBEncryption internal val withEncryption: Boolean
) {
    val writableDatabase: Database
        get() = if (withEncryption) {
            encryptedWritableDb()
        } else {
            databaseOpenHelper.writableDb
        }

    private fun encryptedWritableDb(): Database {
        try {
            if (!upgradeDBFormatStorage.isUpgraded) {
                try {
                    databaseOpenHelper.updateDatabaseFormat(databaseSecretProvider.secret)
                } catch (ex: SQLiteException) {
                    databaseOpenHelper.updateDatabaseFormat(databaseSecretProvider.default)
                }
                upgradeDBFormatStorage.isUpgraded = true
            }
        } catch (e: Exception) {
            upgradeDBFormatStorage.isUpgraded = true
        }

        return try {
            databaseOpenHelper.getEncryptedWritableDb(databaseSecretProvider.secret)
        } catch (ex: SQLiteException) {
            databaseOpenHelper.getEncryptedWritableDb(databaseSecretProvider.default)
        }
    }

    companion object {
        val TAG: String = CoinKeeperOpenHelper::class.java.simpleName
    }

}